codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -161,6 +161,8 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                     }
                     throw cause;
                 }
+            } else if (stream.isTrailersSent() || stream.isHeadersSent() && !endOfStream) {
+                throw new IllegalStateException("Stream " + stream.id() + " sent too many headers EOS: " + endOfStream);
```

Comment:
It is failed in the `catch` block below. The only exception to this control flow is when the stream no longer exists ... we catch early and prevent `lifecycleManager.onError` from being called.